neps/nep-0568.md

```diff
@@ -63,14 +63,45 @@ post-processing, as long as the chain's view reflects a fully resharded state.
   shards post-resharding.
-* Cross-Shard Traffic: Receipts sent to the parent shard may need to be
-  reassigned to one of the child shards.
+* Cross-Shard Traffic: Receipts and buffered receipts sent to the parent shard may
+  need to be reassigned to one of the child shards.
 * Receipt Handling: Delayed, postponed, buffered, and promise-yield receipts
   must be correctly distributed between the child shards.
 * ShardId Semantics: The shard identifiers will become abstract identifiers
   where today they are number in the 0..num_shards range.
+* Congestion Info: CongestionInfo in the chunk header would be recalculated for the child
+  shards at the resharding boundary. Proof must be compatible with Stateless Validation.
 
-### State Storage - Mem Trie
+### State Storage - MemTrie
+
+MemTrie is the in-memory representation of the trie that the runtime uses for all trie accesses. This is kept in sync with the Trie representation in state.
+
+For the purposes of resharding, we need an efficient way to split the MemTrie into two child tries based on the boundary account. This splitting happens at the epoch boundary when the new epoch is expected to have the two child shards. The set of requirements around MemTrie splitting are:
+* MemTrie splitting needs to be "instant", i.e. happen efficiently within the span of one block. The child tries need to be available for the processing of the next block in the new epoch.
+* MemTrie splitting needs to be compatible with stateless validation, i.e. we need to generate a proof that the memtrie split proposed by the chunk producer is correct.
+* The proof generated for splitting the MemTrie needs to be compatible with the limits of the size of state witness that we send to all chunk validators. This prevents us from doing things like iterating through all trie keys for delayed receipts etc.
+
+With ReshardingV3 design, there's no protocol change to the structure of MemTries, however the implementation constraints required us to introduce the concept of a Frozen MemTrie. More details are in the [implementation](#state-storage---memtrie-1) section below.
+
+Based on the requirements above, we came up with an algorithm to efficiently split the parent trie into two child tries. Trie entries can be divided into three categories based on whether the trie keys have an account_id prefix and based on the total number of such trie keys. Splitting of these keys are handled in different ways.
+
+#### TrieKey with AccountID prefix
+
+This category includes most of the trie keys like `TrieKey::Account`, `TrieKey::ContractCode`, `TrieKey::PostponedReceipt`. For these keys, we can efficiently split the trie based on the boundary account trie key. In the example below, "pass" was the split key, note that we only need to read all the intermediate nodes that form a part of the split key and nothing more. The accessed nodes form a part of the state witness. This limits the size of the witness to effectively O(depth) of trie.
```

Comment:
"note that we only need to read all the intermediate nodes that form a part of the split key" 
Is this needed for building the state witness proof? While I understand the idea, it would be great if the purpose can be made more explicit